packages/core-plugin-api/src/analytics/useAnalytics.tsx

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useAnalyticsContext } from './AnalyticsContext';
+import { analyticsApiRef, AnalyticsTracker, useApi } from '../apis';
+import { useRef } from 'react';
+import { Tracker } from './Tracker';
+
+/**
+ * Get a pre-configured analytics tracker.
+ */
+export function useAnalytics(): AnalyticsTracker {
+  const trackerRef = useRef<Tracker | null>(null);
+  const analyticsApi = useApi(analyticsApiRef);
```

Comment:
Regarding the `useTracker` discussion, it's very true that that's a bit of a PITA. Let's go with a mix instead? Keep this and the default implementation, because we don't really want plugins to be overriding API implementations willy-nilly under the hood anyway. Then we can keep the presence of the API optional by using `useApiHolder().get(analyticsApiRef)` instead of `useApi`, which will return `undefined` if there's no implementation.

Reply:
Hmm @Rugvip...  We still end up breaking things in tests downstream because `useApiHolder` expects a versioned API context.

```
    API context is not available

      22 |   const versionedHolder = useVersionedContext<{ 1: ApiHolder }>('api-context');
      23 |   if (!versionedHolder) {
    > 24 |     throw new Error('API context is not available');
         |           ^
      25 |   }
      26 |
      27 |   const apiHolder = versionedHolder.atVersion(1);
```

Is

Reply:
Kinda dirty, but I can basically re-implement `useApiHolder()`:

```diff
diff --git a/packages/core-plugin-api/src/analytics/useAnalytics.tsx b/packages/core-plugin-api/src/analytics/useAnalytics.tsx
index c3a1ae3b9c..306eca9822 100644
--- a/packages/core-plugin-api/src/analytics/useAnalytics.tsx
+++ b/packages/core-plugin-api/src/analytics/useAnalytics.tsx
@@ -15,16 +15,28 @@
  */
 
 import { useAnalyticsContext } from './AnalyticsContext';
-import { analyticsApiRef, AnalyticsTracker, useApi } from '../apis';
+import { analyticsApiRef, AnalyticsTracker, AnalyticsApi, ApiHolder } from '../apis';
 import { useRef } from 'react';
 import { Tracker } from './Tracker';
+import { useVersionedContext } from '@backstage/version-bridge';
+
+function useTracker(): AnalyticsApi {
+  // We effectively reimplement useApiHolder in order to avoid hook rule order
+  // issues so that the analytics API is truly optional in any consuming code.
+  const versionedHolder = useVersionedContext<{ 1: ApiHolder }>('api-context');
+  const apiHolder = versionedHolder?.atVersion(1);
+  return apiHolder?.get(analyticsApiRef) || {
+    // If no analyticsApi is defined, provide a no-op tracker.
+    captureEvent: () => {},
+  };
+};
 
 /**
  * Get a pre-configured analytics tracker.
  */
 export function useAnalytics(): AnalyticsTracker {
   const trackerRef = useRef<Tracker | null>(null);
-  const analyticsApi = useApi(analyticsApiRef);
+  const analyticsApi = useTracker();
   const context = useAnalyticsContext();
 
   function getTracker(): Tracker {
```

Reply:
Ah shiet that's true. Tbh I actually prefer the `try ... catch` at that point, but in a way where we take care to have the same hooks being called regardless of path. My worry with reimplementing the versioned thing is that we would forget to update this site whenever we update the `useApiHolder` implementation

Reply:
Sounds good.  Check it out now?

I'll book some time with you to mob on the version bridge.  I also have a codemod waiting at the ready to get the extension names in place.

Reply:
Sounds good.  Check it out now, @Rugvip?

I'll book some time with you to mob on the version bridge.  I also have a codemod waiting at the ready to get the extension names in place.